core/src/main/scala/cromwell/util/JsonEditor.scala

```diff
@@ -72,21 +73,77 @@ object JsonEditor {
     *
     * @param json json blob with or without "labels" field
-    * @param labels a map of labels one would like to apply to a workflow json
+    * @param databaseLabels a map of workflow IDs to maps of labels one would like to apply to a workflow json
     * @return json with labels merged in.  Any prior non-object "labels" field will be overwritten and any object fields will be merged together and - again - any existing values overwritten.
     */
-  def augmentLabels(json: Json, labels: Map[String, String]): Json = {
-    val newData: Json = Json.fromFields(labels.safeMapValues(Json.fromString))
-    val newObj: Json = Json.fromFields(List(("labels", newData)))
-    //in the event of a key clash, the values in "newObj" will be favored over "json"
-    json deepMerge newObj
-  }
+  def updateLabels(json: Json, databaseLabels: Map[WorkflowId, Map[String, String]]): Json = {
+    val subWorkflowMetadataKey = "subWorkflowMetadata"
+
+    def doUpdateWorkflow(workflowJson: Json): Json = {
+      val id: String = workflowJson.rootWorkflowId.
+        getOrElse(throw new RuntimeException(s"did not find workflow id in ${workflowJson.printWith(Printer.spaces2)}")).toString
+
+      // Look for an optional JsonObject keyed by "calls".
+      val callsObject: Option[JsonObject] = for {
+        wo <- workflowJson.asObject
+        callsJson <- wo("calls")
+        co <- callsJson.asObject
+      } yield co
+
+      val workflowWithUpdatedCalls: Json = callsObject match {
+        // If there were no calls just return the workflow JSON unmodified.
+        case None => workflowJson
+        case Some(calls) =>
+          val updatedCallsObject = calls.mapValues {
+            // The Json (a JSON array, really) corresponding to the array of call objects for a call name.
+            callValue: Json =>
+              // The object above converted to a List[Json].
+              val callArray: List[Json] = callValue.asArray.toList flatMap { _.toList }
 
-  def removeSubworkflowData(json: Json): Json = excludeJson(json, NonEmptyList.of("subWorkflowMetadata"))
+              val updatedCallArray = callArray map { callJson =>
+                // If there is no subworkflow object this will be None.
+                val callAndSubworkflowObjects: Option[(JsonObject, JsonObject)] = for {
+                  co <- callJson.asObject
+                  sub <- co(subWorkflowMetadataKey)
+                  so <- sub.asObject
+                } yield (co, so)
+
+                callAndSubworkflowObjects match {
+                  case None => callJson
+                  case Some((callObject, subworkflowObject)) =>
+                    // If the call contains a subWorkflowMetadata key, return a copy of the call with
+                    // its subworkflowMetadata updated via a recursive call to `doUpdateWorkflow`.
+                    val updatedSubworkflow = doUpdateWorkflow(Json.fromJsonObject(subworkflowObject))
```

Comment:
TOL: Line 122 maps this to an object and then you immediately revert it to a `Json` here... is that deliberate (eg in case it wasn't an object but some other type)?

Reply:
Yes but that doesn't mean I couldn't have returned the `Json` bit I really wanted and avoided this unnecessary reversion...